File: javascript/grains/grains.js
import BigInt from './lib/big-integer';

export class Grains {
  square(square) {
    return BigInt(2).pow(square - 1).toString();
  }

  total() {
    let total = BigInt(0);
    for (let i = 1; i <= 64; i++) {
      total = total.add(BigInt(2).pow(i - 1));
    }
    return total.toString();
  }

  // this is actually slower using the finite geometric series formula
  // sum from i = 1..n of 2^i = 2 * (2^n - 1)
  // total() { 
  //   return BigInt(2).pow(63).minus(1).times(2).add(1).toString();
  // }
}

File: python/grains/grains.py
def square(number):
    if not 1 <= number <= 64:
        raise ValueError('Out of range')
    return 2 ** (number - 1)


def total():
    # return sum(2**(n - 1) for n in range(1, 65))
    return 2**64 - 1
